src/basic/MAT_MAT_SHARED-OMPTarget.cpp

```diff
@@ -0,0 +1,44 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
+// and RAJA Performance Suite project contributors.
+// See the RAJAPerf/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "MAT_MAT_SHARED.hpp"
+
+#include "RAJA/RAJA.hpp"
+
+#if defined(RAJA_ENABLE_TARGET_OPENMP)
+
+#include "common/OpenMPTargetDataUtils.hpp"
+
+#include <iostream>
+
+namespace rajaperf {
+namespace basic {
+
+void MAT_MAT_SHARED::runOpenMPTargetVariant(VariantID vid) {
+
+  MAT_MAT_SHARED_DATA_SETUP;
+
+  if (vid == Base_OpenMPTarget) {
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+    }
+    stopTimer();
```

Comment:
Should this just be empty, or even not here at all?

Reply:
Is there no way to use device shared memory in OpenMP target offload?

Reply:
Chatting with @rchen20 it might be a feature in OpenMP 5.1. It would be really cool if we can get that going. @trws what do you think? Has anyone tried it here?

Reply:
It's technically supported farther back than that, 5.0 IIRC.  There is some argument about which forms are meant to work and which aren't, which is not great, but the one that's agreed to work by basically everyone (and I think is implemented in at least clang) looks like this:

```c
double my_shared_mem[size];
#pragma omp target teams distribute allocate(omp_cgroup_mem_alloc: my_shared_mem) private(my_shared_mem)
#pragma omp parallel for
<loop> // my_shared_mem is private to the team, separate for each contention group, so it should be shared/local memory
```

Constant memory should also be possible like this, but I have yet to test it anywhere:

```c
double my_const_mem[size];
#pragma omp target teams distribute allocate(omp_const_mem_alloc: my_shared_mem) firstprivate(my_const_mem)
#pragma omp parallel for
<loop> // my_shared_mem is private to the team, separate for each contention group, so it should be shared/local memory
```

Note this one has firstprivate, that's the only way to write into a constant memory segment allocated by that openmp memory allocator right now.

Reply:
I'd note that both of the above, as well as the example below, are required to give you the *semantics* of shared memory on a gpu on any device where they execute.  Whether they actually use shared memory on a given GPU is QoI, so having a benchmark sitting around we can use to show when appropriate memories aren't being used might be really good.

Just using private at the teams level, or declaring a variable at that level, should also work.

Reply:
@trws in regards to the last comment, when you say declaring the variable at the teams level would it look something like this  (In the code example I'm also thinking about how to implement this in RAJA teams): 

```
#pragma omp target ( this could be what the RAJA::launch method sets up) 
{

  #pragma omp teams distribute (RAJA team loop)
   for(e=0; e<ne_teams; ++e) { 

     double  sm_mem [ ]  // memory exclusive to each team

    #pragma parallel for (RAJA thread loop)
    for(t=0; t<threads; ++t) {
      sm_mem[t] = data; //sm_mem is available to all threads within a team. 
    }

  }
}
```

Reply:
Yup, that's basically it.  Depending on the pattern it may be problematic to have it inside the distribute but that should *work*.  It likely will not perform well, compilers generally don't make that into shared memory (at least not yet), but it should work.

The form that should make it work *well* for this kind of case ought to look about like this:

```c
#pragma omp target ( this could be what the RAJA::launch method sets up) 
{

  #pragma omp teams distribute (RAJA team loop)
   for(e=0; e<ne_teams; ++e) { 

     double  sm_mem [ ]  // memory exclusive to each team
     #pragma omp allocate(omp_cgroup_mem_alloc:sm_mem)
    #pragma parallel for (RAJA thread loop)
    for(t=0; t<threads; ++t) {
      sm_mem[t] = t+1.5; //sm_mem is available to all threads within a team. 
    }

  }
}
```

The concern some people have right now is whether that's guaranteed to work the way we intended to by the text in the spec.  It's literally on the docket for discussion in the F2F next week.  I would think it should work, but whether implementers agree with me is an open question.

Reply:
Cool! I think adding an omp target variant in would be great. I am partial to the version which does not require the `omp allocate` directive. The downside of the allocate directive is in the case where were have many arrays we would like to store in shared memory,  it can then get kind of verbose.  

How would you guys feel about adding the OMP target variant after we add the support in teams? As mentioned it would be useful for engaging with the compiler folks and illustrate patterns/styles we use.

Reply:
It would be worth doing.  I'd recommend using a macro or something so that the allocate directive can be included or not though, I strongly suspect that the teams level declaration without it will become a big global memory allocation and perform *very* badly.  If that's something we want compilers to support better than that, it's a discussion we can have with them, but being able to compare the two would probably help a lot.